sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1155,17 +1324,24 @@ def logging_api_service_descriptor(self):
         url='localhost:%s' % self.logging_port)
 
   def close(self):
-    self.control_handler.done()
-    self.data_plane_handler.close()
-    to_wait = [
-        self.control_server.stop(self._DEFAULT_SHUTDOWN_TIMEOUT_SECS),
-        self.data_server.stop(self._DEFAULT_SHUTDOWN_TIMEOUT_SECS),
-        self.state_server.stop(self._DEFAULT_SHUTDOWN_TIMEOUT_SECS),
-        self.logging_server.stop(self._DEFAULT_SHUTDOWN_TIMEOUT_SECS)
-    ]
-    for w in to_wait:
-      w.wait()
-    super(GrpcWorkerHandler, self).close()
+    # Wait all workers are done and close GRPC server.
+    WorkerHandler._worker_cnt -= 1
+    if WorkerHandler._worker_cnt == 0:
```

Comment:
Seems like this logic belongs in GrpcServer(). 

Maybe have a static GrpcServer.get_instance() method that does the caching and incrementing (rather than a static member on GrpcWorkerHandler) together with a release_instance() method that's called in GrpcWorkerHandler.close() (which may cause it to shut down). Right now one class is responsible for incrementing, another for decrementing, making it harder to follow.